### PR TITLE
PR: improving macro with adding attributes properties for struct

### PR DESCRIPTION
**Pull request by fatihaziz**
Here my update regarding to support:
1. Override generated attribute macro for Struct
2. Handle attribute macro for _new fields_

Thank you, best regard

### CODE REVIEW

**Comment by douggynix** [comment]
By the way, what type of code formatter are you using? i think you're using a special code formatting settings in your IDE that changes all of the source code files formatting. I was using cargo default code formatter. 
```shell
cargo fmt --verbose
```

**Review by douggynix — CHANGES_REQUESTED** [review]
Thanks for your contribution for the macro attributes per mapper entry field that targets generated structs :).

All of the other changes look good to me. just submit those requested changes in the comments and I'll go on after that to merge the PR.

I'll open up another PR as well after that to proceed with "field attribute" macro as you already lay the foundation codes to parse and generate macro attributes.

**Comment by fatihaziz** [comment]
> By the way, what type of code formatter are you using? i think you're using a special code formatting settings in your IDE that changes all of the source code files formatting. I was using cargo default code formatter.
> 
> ```shell
> cargo fmt --verbose
> ```

yes i use the same,, just i put wordwrap 100 here so it might different from yours

**Comment by fatihaziz** [comment]
> Thanks for your contribution for the macro attributes per mapper entry field that targets generated structs :).
> 
> All of the other changes look good to me. just submit those requested changes in the comments and I'll go on after that to merge the PR.
> 
> Do you consider to proceed as well with "field attribute" macros as you already lay the foundation codes to parse and generate macro attributes.
> 
> if so, the field attributes should be part of the map attribute such as the variants would look like this as before:
> 
> ```rust
> // the order of the elements in the tuple doesn't matter, only the type matters as they're different.
> We have a String Literal type, a Boolean literal type, and an Array of Literal type for the macro attributes
> map=([   
>      ( "field1:fieldA",  true,  [ "attrA", "attrB" ] ) ,  ( "field3:fieldB", [ "attrE" ] ),  ( "field2", [ "attrC", "AttrD" ] ), 
>  ])
> ```

i love to also adding that functionality on the _map_ attribute also, let me check it

File: examples/dto_example.rs [hunk 1]
```diff
@@ -6,34 +6,34 @@ use std::str::FromStr;
...
+  s1.to_owned() + " " + s2
 }
 
 #[derive(DtoMapper, Debug, Default, Clone)]
 #[mapper( dto="LoginDto"  , map=[ ("username:login",true) , ("password",true)] , derive=(Debug, Clone, PartialEq) )]
 #[mapper( dto="ProfileDto" , ignore=["password"]  , derive=(Debug, Clone, PartialEq) )]
 #[mapper( dto="PersonDto" , no_builder=true , map=[ ("firstname",true), ("lastname",true), ("email",false) ]  )]
 #[mapper( dto="CustomDto" , no_builder=true , map=[ ("email",false) ] , derive=(Debug, Clone) ,
-new_fields=[( "name: String", "concat_str( self.firstname.as_str(), self.lastname.as_str() )" )]
+  new_fields=[( "name: String", "concat_str( self.firstname.as_str(), self.lastname.as_str() )" )]
```
[the remark this answers is not in the document]
Comment (fatihaziz) [reply examples/dto_example.rs:17]: sure will add that also

**Comment by douggynix** [comment]
> > Thanks for your contribution for the macro attributes per mapper entry field that targets generated structs :).
> > All of the other changes look good to me. just submit those requested changes in the comments and I'll go on after that to merge the PR.
> > Do you consider to proceed as well with "field attribute" macros as you already lay the foundation codes to parse and generate macro attributes.
> > if so, the field attributes should be part of the map attribute such as the variants would look like this as before:
> > ```rust
> > // the order of the elements in the tuple doesn't matter, only the type matters as they're different.
> > We have a String Literal type, a Boolean literal type, and an Array of Literal type for the macro attributes
> > map=([   
> >      ( "field1:fieldA",  true,  [ "attrA", "attrB" ] ) ,  ( "field3:fieldB", [ "attrE" ] ),  ( "field2", [ "attrC", "AttrD" ] ), 
> >  ])
> > ```
> 
> i love to also adding that functionality on the _map_ attribute also, let me check it

Yeah, it would be a great idea doing it in the map. I already have codes that parse the Boolean and the "String Literal", you can just add another if case that parses the attributes here:

**Comment by douggynix** [comment]
> > By the way, what type of code formatter are you using? i think you're using a special code formatting settings in your IDE that changes all of the source code files formatting. I was using cargo default code formatter.
> > ```shell
> > cargo fmt --verbose
> > ```
> 
> yes i use the same,, just i put wordwrap 100 here so it might different from yours, i will disable it for this project only, dont worry

Please, you can disable it so that we can ease code review and eliminates noise for formatting.
Thanks

File: tests/test_dto_creation.rs [hunk 2]
```diff
@@ -1,136 +1,167 @@
...
+    dto="CustomDtoWithAttribute" ,
+    no_builder=true ,
+    map=[ ("email",false) ],
+    derive=(Debug, Clone, Serialize, Deserialize),
+    new_fields=[(
+        "name: String",
+        "concat_str( self.firstname.as_str(), self.lastname.as_str() )",
+        ["#[serde(rename = \"renamed_name\")]"],
+    )],
+    macro_attr=["serde(rename_all = \"UPPERCASE\")"]
```
Comment (douggynix) [inline tests/test_dto_creation.rs:32]: if we can have macro attributes for existing fields as well with the use of map keyword already in use.

**Review by douggynix — CHANGES_REQUESTED** [review]

File: src/mapper_entry.rs [hunk 3]
```diff
@@ -82,276 +91,390 @@ const IGNORE: &'static str = "ignore";
...
-            panic!("Need to specify a type for the fieldname after `:`");
+        // Process the field when we have all provided arguments
+        if total_passed_args - 1 == position {
+          if let (Some(field_decl), Some(field_value)) =
+            (&field_data[0], &field_data[1])
+          {
+            if let Some(colon_position) = field_decl.find(":") {
+              // eprintln!("field_decl={}", field_decl);
+              // eprintln!("colon={}", colon_position);
+              Self::insert_next_field_value(
```
Comment (douggynix) [inline src/mapper_entry.rs:394]: Could you replace this line by this one for code readability please:
```rust
              let new_field_attr = match attributes.is_empty(){
                true => None,
                false => Some(attributes.clone())
              };
              
              Self::insert_next_field_value(
                &mut vec_tuple,
                field_value.clone(),
                field_decl,
                &colon_position,
                new_field_attr
              );
```

**Review by douggynix — CHANGES_REQUESTED** [review]

**Comment by douggynix** [comment]
Any update on this PR. Or could I provide any coding help with my suggestions above. 
Thanks

**Comment by douggynix** [comment]
@fatihaziz As I have no feedback from you for the code review. I am going to merge the code onto the main branch and proceed with adding the changes highlighted in the review. I will open up another PR after merge so that you can see the related changes added.

Thanks a lot for your help and contribution to this code base.

**Comment by fatihaziz** [comment]
@douggynix I'm very you don't hear update from free, i got very busy on my job.
thank you for your work, its work like i imagine!

**Pull request merged**